src/providers/cryptoauthlib/key_slot.rs

```diff
@@ -70,21 +73,32 @@ impl AteccKeySlot {
         Ok(())
     }
 
-    fn is_key_type_ok(&self, key_attr: &Attributes) -> bool {
+    fn is_key_type_ok(&self, slot: u8, key_attr: &Attributes) -> bool {
         match key_attr.key_type {
             Type::RawData => self.config.key_type == rust_cryptoauthlib::KeyType::ShaOrText,
             Type::Hmac => !self.config.no_mac,
             Type::Aes => self.config.key_type == rust_cryptoauthlib::KeyType::Aes,
             Type::EccKeyPair {
                 curve_family: EccFamily::SecpR1,
+            } => {
+                // P256 private key has 256 bits (32 bytes).
+                // Only private key is stored - public one can be computed when needed.
+                // The private key can onlly be stored encrypted and the encryption key must be set,
+                // see set_write_encryption_key() call in new().
+                key_attr.bits == 256
+                    && self.config.key_type == rust_cryptoauthlib::KeyType::P256EccKey
+                    && self.config.write_config == rust_cryptoauthlib::WriteConfig::Encrypt
+                    && self.config.ecc_key_attr.is_private
+                    && self.config.is_secret
             }
-            | Type::EccPublicKey {
+            Type::EccPublicKey {
                 curve_family: EccFamily::SecpR1,
             } => {
-                // There may be a problem here: P256 private key has 256 bits (32 bytes),
-                // but the uncompressed public key is 512 bits (64 bytes)
-                key_attr.bits == 256
+                // The uncompressed public key is 512 bits (64 bytes).
+                // First few (7) slots are too short for ECC public key.
+                key_attr.bits == 512
```

Comment:
Note from the [PsaImportKey](https://parallaxsecond.github.io/parsec-book/parsec_client/operations/psa_import_key.html) contract:

> The key size is always determined from the data buffer. If the key size in attributes is nonzero, it must be equal to the size from data.

Explained in a bit more detail in the [PSA API](https://armmbed.github.io/mbed-crypto/html/api/keys/management.html#psa_import_key).

What I mean is that the `bits` field does not correspond to the size of the data buffer that is imported but the size of the key, and for ECC keys that is the bit size associated with the curve. See the expected format [here](https://parallaxsecond.github.io/parsec-book/parsec_client/operations/psa_export_public_key.html#description).

So that if a ECC public key of a curve `secp256r1` is imported, then its `bits` field should still be 256 even if the data buffer is larger than that (512 in this case?).

Reply:
Let me re-phrase:
1. if `bits` is equal 0, then only the appropriate contents of `data` determine key size (a number of bits).
2. if `bits` is larger than 0, then it must match the appropriate contents of `data` BUT for cryptoauthlib provider only 256 is acceptable, be it for private or public key.

Is that right?

Reply:
Yes that's correct! So ideally you should parse `data` to calculate the key size and if `bits` is larger than 0, make sure that the two values are equal. Since only 256 is supproted I guess you would need to check that `bits` the the key size are equal to 256.

There aren't any example of parsing ECC keys in Parsec at the moment. For the ECC key pairs, the format is explained  [here](https://parallaxsecond.github.io/parsec-book/parsec_client/operations/psa_export_key.html#description) and is the ASN.1 type `ECPrivateKey`:

```
   ECPrivateKey ::= SEQUENCE {
     version        INTEGER { ecPrivkeyVer1(1) } (ecPrivkeyVer1),
     privateKey     OCTET STRING,
     parameters [0] ECParameters {{ NamedCurve }} OPTIONAL,
     publicKey  [1] BIT STRING OPTIONAL
   }
```

We are using the `picky` crates to do ASN.1 parsing/serialization and before we were also defining our own ASN.1 type, like for RSA keys. See #201 for example. Maybe we can convince @CBenoit to also add support for `ECPrivateKey` in `picky-asn1-x509` or another crate?

[This](https://parallaxsecond.github.io/parsec-book/parsec_client/operations/psa_export_public_key.html#description) explains the representation of the ECC public key, which is not an ASN.1 structure and will need manual parsing I think.

Reply:
Thanks.  I did not find any converter of a binary private key into this format. I will try harder.

Reply:
Hi, just passing by! ECC is on my wishlist for `picky` (I think ASN.1 types would live in `picky-asn1-x509` as well), but I have yet to find the time to actually look into it and implement it (or someone to open a PR).

Reply:
Thanks a lot @CBenoit! I guess we can try by defining the new `ECPrivateKey` type ourselves and then make a PR to `picky-asn1-x509` if it works well.
@RobertDrazkowskiGL I am happy to help with that!

Reply:
Corrected. About parsing ASN.1 - at the moment we are doing it "manually", i.e. we are calling openssl functions for doing the hard job for us. But of course we plan to submit PRs to picky-asn1, hopefully soon.